include/seastar/coroutine/async_generator.hh

```diff
@@ -0,0 +1,421 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+#include <cassert>
+#include <coroutine>
+#include <exception>
+#include <iterator>
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <seastar/core/future.hh>
+
+// async_generator concept is heavily inspired by P2502R2
+// (https://wg21.link/P2502R2), a proposal accepted into C++23. P2502R2
+// introduced std::generator, which provides a synchronous coroutine
+// mechanism for generating ranges. in contrast, async_generator offers
+// asynchronous generation of element sequences.
+namespace seastar::coroutine::experimental {
+
+template<typename Ref, typename Value = void>
+class async_generator;
+
+namespace internal {
+
+template <typename Yielded> class next_awaiter;
+
+template <typename Yielded>
+class async_generator_promise_base : public seastar::task {
+protected:
+    std::add_pointer_t<Yielded> _value = nullptr;
+
+protected:
+    std::exception_ptr _exception;
+    std::coroutine_handle<> _consumer;
+    task* _waiting_task = nullptr;
+
+    class yield_awaiter final {
+        async_generator_promise_base* _promise;
+        std::coroutine_handle<> _consumer;
```

Comment:
Do we know that the consumer is a coroutine?

Reply:
hmm, i am confused. if the consumer is not a coroutine, how come can it `co_await` a generator? in this case, the consumer needs to `co_await` for the iterator .

Reply:
It's just so confusing. The generator itself is a coroutine, calling co_yield, but so is its caller, calling co_await.

Reply:
But no, it can't be true. The caller doesn't know that the generator is a coroutine. It should be possible to use a generator not within a coroutine.

Reply:
The caller of the coroutine receives a seastar::coroutine::async_generator object. It doesn't have to co_await it. Or maybe, that's the only operations it exposes (operator co_await).

I think we should allow non-coroutines to interact with generators. Perhaps via a future\<std::optional\<T>> operator()(), like our mutation_readers.

Meanwhile please add comments to denote what the use case is here. It's just so overwhelming.